core/server/src/main/java/alluxio/worker/netty/DataServerWriteHandler.java

```diff
@@ -0,0 +1,337 @@
+      }
+      mPosToQueue += buf.readableBytes();
+      mPackets.offer(buf);
+      if (!mPacketWriterActive) {
+        mPacketWriterExecutor.submit(new PacketWriter(ctx));
+        mPacketWriterActive = true;
+      }
+      if (tooManyPacketsInFlight()) {
+        ctx.channel().config().setAutoRead(false);
+      }
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+    LOG.error("Failed to write block " + (mRequest == null ? -1 : mRequest.mId) + ".", cause);
+    replyError(ctx.channel(), Protocol.Status.Code.INTERNAL, "", cause);
+  }
+
+  @Override
+  public void channelUnregistered(ChannelHandlerContext ctx) {
+    try {
+      reset();
+    } catch (IOException e) {
+      LOG.warn("Failed to reset the write request inside channelUnregistered.");
+    }
+    ctx.fireChannelUnregistered();
+  }
+
+  /**
+   * @return true if there are too many packets in flight
+   */
+  private boolean tooManyPacketsInFlight() {
+    return mPackets.size() >= MAX_PACKETS_IN_FLIGHT;
+  }
+
+  /**
+   * Validates the block write request.
+   *
+   * @param msg the block write request
+   * @return empty string if the request valid
+   */
+  private String validateRequest(RPCProtoMessage msg) {
+    Protocol.WriteRequest request = (Protocol.WriteRequest) msg.getMessage();
+    if (request.getId() != mRequest.mId) {
+      return "The Ids do not match.";
+    }
+    if (request.getOffset() != mPosToQueue) {
+      return String
+          .format("Offsets do not match [received: %d, expected: %d].", request.getOffset(),
+              mPosToQueue);
+    }
+    return "";
+  }
+
+  /**
+   * Writes an error block write response to the channel and closes the channel after that.
+   *
+   * @param channel the channel
+   */
+  private void replyError(Channel channel, Protocol.Status.Code code, String message, Throwable e) {
+    channel.writeAndFlush(RPCProtoMessage.createResponse(code, message, e, null))
+        .addListener(ChannelFutureListener.CLOSE);
+  }
+
+  /**
+   * Writes a response to signify the success of the block write. Also resets the channel.
+   *
+   * @param channel the channel
+   */
+  private void replySuccess(Channel channel) {
+    channel.writeAndFlush(RPCProtoMessage.createOkResponse(null))
+        .addListeners(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE, new ChannelFutureListener() {
+          @Override
+          public void operationComplete(ChannelFuture future) throws Exception {
+            reset();
+          }
+        });
+    channel.config().setAutoRead(true);
+    channel.read();
+  }
+
+  private void reset() throws IOException {
+    if (mRequest != null) {
+      mRequest.close();
+      mRequest = null;
+    }
+
+    mPosToQueue = 0;
+    mPosToWrite = 0;
+
+    try {
+      mLock.lock();
+      for (ByteBuf buf : mPackets) {
+        ReferenceCountUtil.release(buf);
+      }
+      mPacketWriterActive = false;
```

Comment:
This doesn't actually do anything to stop the writer thread. Could it be possible that the thread still is still running?